File: scraper-service/scraper/jsonld.py
"""Parses schema.org Restaurant/Menu JSON-LD (https://schema.org/Menu) — many
restaurant websites embed this for Google's rich-snippet requirements, which
makes it by far the most reliable extraction path when it's present. Always
try this before falling back to heuristic DOM scraping."""

from __future__ import annotations

import json
from typing import Any, Iterable

from .normalize import build_category, build_menu_item, price_to_cents


def _iter_jsonld_blocks(html_script_texts: Iterable[str]) -> Iterable[dict[str, Any]]:
    for raw in html_script_texts:
        try:
            data = json.loads(raw)
        except (json.JSONDecodeError, TypeError):
            continue
        if isinstance(data, list):
            yield from (d for d in data if isinstance(d, dict))
        elif isinstance(data, dict):
            if "@graph" in data and isinstance(data["@graph"], list):
                yield from (d for d in data["@graph"] if isinstance(d, dict))
            else:
                yield data


def _find_menu_node(node: Any) -> dict[str, Any] | None:
    if not isinstance(node, dict):
        return None
    node_type = node.get("@type")
    types = node_type if isinstance(node_type, list) else [node_type]
    if "Menu" in types:
        return node
    if "Restaurant" in types or "FoodEstablishment" in types:
        has_menu = node.get("hasMenu")
        if isinstance(has_menu, dict):
            return has_menu
        if isinstance(has_menu, list) and has_menu:
            return has_menu[0]
    return None


def _menu_item_to_dict(item: dict[str, Any]) -> dict[str, Any] | None:
    name = item.get("name")
    if not name:
        return None
    offers = item.get("offers")
    price = None
    if isinstance(offers, dict):
        price = offers.get("price")
    elif isinstance(offers, list) and offers:
        price = offers[0].get("price")
    return build_menu_item(str(name), price_to_cents(price), item.get("description"))


def extract_menu_from_jsonld(html_script_texts: list[str]) -> dict[str, Any] | None:
    """Returns a ProspectMenu-shaped dict, or None if no schema.org menu was found."""
    for block in _iter_jsonld_blocks(html_script_texts):
        menu_node = _find_menu_node(block)
        if not menu_node:
            continue

        categories: list[dict[str, Any]] = []
        sections = menu_node.get("hasMenuSection")
        sections = sections if isinstance(sections, list) else ([sections] if sections else [])

        for section in sections:
            if not isinstance(section, dict):
                continue
            section_items = section.get("hasMenuItem")
            section_items = section_items if isinstance(section_items, list) else ([section_items] if section_items else [])
            items = [d for d in (_menu_item_to_dict(i) for i in section_items if isinstance(i, dict)) if d]
            if items:
                categories.append(build_category(str(section.get("name") or "Menu"), items))

        # Some sites list menu items directly on the Menu node with no sections.
        if not categories:
            direct_items = menu_node.get("hasMenuItem")
            direct_items = direct_items if isinstance(direct_items, list) else ([direct_items] if direct_items else [])
            items = [d for d in (_menu_item_to_dict(i) for i in direct_items if isinstance(i, dict)) if d]
            if items:
                categories.append(build_category("Menu", items))

        if categories:
            return {"categories": categories}

    return None

File: lib/prospects/scrape/jsonld.ts
import "server-only";
import * as cheerio from "cheerio";
import { nextId, priceToCents } from "../menu-text-helpers";
import type { ProspectMenu, ProspectMenuCategory, ProspectMenuItem } from "../types";

/**
 * Parses schema.org Restaurant/Menu JSON-LD (https://schema.org/Menu) out of a page's
 * raw HTML — many restaurant websites embed this for Google's rich-snippet requirements,
 * which makes it by far the most reliable extraction path when it's present, and it
 * requires no JS rendering since it's always server-rendered for search crawlers.
 */
export function extractMenuFromJsonLd(html: string): ProspectMenu | null {
  const $ = cheerio.load(html);
  const blocks: unknown[] = [];

  $('script[type="application/ld+json"]').each((_, el) => {
    const raw = $(el).contents().text();
    try {
      const parsed = JSON.parse(raw);
      if (Array.isArray(parsed)) blocks.push(...parsed);
      else if (parsed && typeof parsed === "object") {
        const graph = (parsed as Record<string, unknown>)["@graph"];
        if (Array.isArray(graph)) blocks.push(...graph);
        else blocks.push(parsed);
      }
    } catch {
      // Malformed JSON-LD on the page — skip it, not our problem to fix.
    }
  });

  for (const block of blocks) {
    const menuNode = findMenuNode(block);
    if (!menuNode) continue;
    const menu = buildMenuFromNode(menuNode);
    if (menu.categories.length > 0) return menu;
  }

  return null;
}

function asArray(value: unknown): unknown[] {
  if (Array.isArray(value)) return value;
  if (value === null || value === undefined) return [];
  return [value];
}

function findMenuNode(node: unknown): Record<string, unknown> | null {
  if (!node || typeof node !== "object") return null;
  const obj = node as Record<string, unknown>;
  const types = asArray(obj["@type"]);

  if (types.includes("Menu")) return obj;
  if (types.includes("Restaurant") || types.includes("FoodEstablishment")) {
    const hasMenu = asArray(obj.hasMenu)[0];
    if (hasMenu && typeof hasMenu === "object") return hasMenu as Record<string, unknown>;
  }
  return null;
}

function itemFromNode(node: unknown): ProspectMenuItem | null {
  if (!node || typeof node !== "object") return null;
  const obj = node as Record<string, unknown>;
  const name = obj.name;
  if (typeof name !== "string" || !name.trim()) return null;

  const offers = asArray(obj.offers)[0] as Record<string, unknown> | undefined;
  const price = offers?.price;

  return {
    id: nextId("item"),
    name: name.trim(),
    description: typeof obj.description === "string" ? obj.description.trim() : undefined,
    priceCents: priceToCents(typeof price === "string" || typeof price === "number" ? price : null) ?? 0,
    inStock: true,
    dietaryTags: [],
    modifierGroups: [],
  };
}

function buildMenuFromNode(menuNode: Record<string, unknown>): ProspectMenu {
  const categories: ProspectMenuCategory[] = [];
  const sections = asArray(menuNode.hasMenuSection);

  for (const section of sections) {
    if (!section || typeof section !== "object") continue;
    const sectionObj = section as Record<string, unknown>;
    const items = asArray(sectionObj.hasMenuItem)
      .map(itemFromNode)
      .filter((i): i is ProspectMenuItem => i !== null);
    if (items.length > 0) {
      categories.push({
        id: nextId("cat"),
        name: typeof sectionObj.name === "string" ? sectionObj.name : "Menu",
        items,
      });
    }
  }

  if (categories.length === 0) {
    const directItems = asArray(menuNode.hasMenuItem)
      .map(itemFromNode)
      .filter((i): i is ProspectMenuItem => i !== null);
    if (directItems.length > 0) {
      categories.push({ id: nextId("cat"), name: "Menu", items: directItems });
    }
  }

  return { categories };
}
